src/msssim.jl

```diff
@@ -0,0 +1,84 @@
+
+!!! info
+
+    MSSSIM, Multi-scale SSIM is defined only for gray images. RGB images are treated as 3d Gray
+    images. General `Color3` images are converted to RGB images first, you could
+    manually expand them using `channelview` if you don't want them converted to
+    RGB first.
+```
+
+# Reference
+
+[1] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2004). Image quality assessment: from error visibility to structural similarity. _IEEE transactions on image processing_, 13(4), 600-612.
+
+"""
+struct MSSSIM{N} <: FullReferenceIQI
+    kernel::AbstractArray{<:Real}
+    W::NTuple{N, <:Tuple} # number of scales inferred from length
+    function MSSSIM(kernel, W)
+        ndims(kernel) == 1 || throw(ArgumentError("only 1-d kernel is valid"))
+        issymetric(kernel) || @warn "MSSSIM kernel is assumed to be symmetric"
+        all(length.(W) .== 3) || throw(ArgumentError("(α, β, γ) required for all scales, instead it's $(W)"))
+        all([ all(x .>= 0) for x in W ]) || throw(ArgumentError("α, β, γ should be non-negative for all scales, instead it's $(W)"))
+        new{length(W)}(kernel, W)
+    end
+end
+
+# Weights for α, β, γ in [1]
+const MSSSIM_W = (0.0448, 0.2856, 0.3001, 0.2363, 0.1333) # α, β, γ for scales 1 through 5
+
+MSSSIM(kernel, W::NTuple{N, <:Real}) where N = MSSSIM(kernel, map(x->(x, x, x), W)) # shorthand for αᵢ=βᵢ=γᵢ for all scales i
+
+# USE SSIM_KERNEL as default kernel
+MSSSIM(kernel=SSIM_KERNEL) = MSSSIM(kernel, MSSSIM_W)
+
+(iqi::MSSSIM)(x, ref) = _msssim_map(iqi, x, ref)
+assess_msssim(x, ref) = MSSSIM()(x, ref)
+
+const DOWNSAMPLE_FILTER = ones(2, 2)./4 # as per author's implementaion
+
+# SSIM does not allow for user specifying peakval and K, so we don't allow it here either
+function _msssim_map(iqi::MSSSIM, x::GenericImage, ref::GenericImage)
+    if size(x) ≠ size(ref)
+        err = ArgumentError("images should be the same size, instead they're $(size(x))-$(size(ref))")
+        throw(err)
+    end
+
+    level = length(iqi.W) # number of scales
+    T = promote_type(float(eltype(ref)), float(eltype(x)))
+    x = of_eltype(T, x)
+    ref = of_eltype(T, ref)
+
+    # check if no. of levels are >= 1
+    level < 1 && throw(ArgumentError("MS-SSIM needs at least one scale"))
+
+    # check if weights are valid - as per authors implementaion
+    sum(sum.(iqi.W)) == 0 && throw(ArgumentError("MS-SSIM must have at least one weight > 0"))
+
+    # downsampling window
+    window = kernelfactors(Tuple(repeated(DOWNSAMPLE_FILTER, ndims(ref))))
+
+    mean_cs = []
+    for i in 1:level-1
+        cs = SSIM(iqi.kernel, (zero(typeof(iqi.W[i][1])), iqi.W[i][2], iqi.W[i][3]))(x, ref)
+        append!(mean_cs, cs)
+
+        imfilter(x, window, "symmetric")
+        imfilter(ref, window, "symmetric")
+
+        # downsampling
+        x = x[ntuple(i->first(axes(x, i)) : 2 : last(axes(x, i)), ndims(x))...]
+        ref = ref[ntuple(i->first(axes(ref, i)) : 2 : last(axes(ref, i)), ndims(ref))...]
```

Comment:
> Correct me if I'm wrong on this, but because channelview (used in _ssim_map under the hood for colored images) does not copy data but only reinterprets a block of memory, repeated calls to it (per level of msssim) shouldn't cause performance issues.

My best guess is it's okay here; usually, if there are only a few calls, it's generally faster as memory copying is more expensive than in-contiguous indexing several times. But if there're too many calls, things might change.

Note that `x = x[indices...]` does copy data, so it's worth a benchmark to check if the view version (`xi = @view x[indices...]`) is faster.

Reply:
_Ran some tests using `BenchmarkTools`' `@benchmark` macro:_

### Grey 512x512
#### Without Views<sup>1</sup>
Images used: `lena_gray_512` and `cameraman` from `TestImages`.
```
memory estimate:  76.20 MiB
allocs estimate:  7607
--------------
minimum time:     76.668 ms (2.71% GC)
median time:      79.206 ms (5.14% GC)
mean time:        80.277 ms (3.96% GC)
maximum time:     90.304 ms (2.30% GC)
--------------
samples:          63
evals/sample:     1
```

#### With Views<sup>2</sup>
```
memory estimate:  75.53 MiB new
allocs estimate:  7619
--------------
minimum time:     77.237 ms (2.63% GC)
median time:      81.212 ms (4.15% GC)
mean time:        81.265 ms (4.20% GC)
maximum time:     88.129 ms (5.41% GC)
--------------
samples:          62
evals/sample:     1
```

### RGB 512x512
Images used: `lena_color_512` and `mandril_color` from `TestImages`.
#### Without Views<sup>1</sup>
```
memory estimate:  383.39 MiB
allocs estimate:  118481
--------------
minimum time:     470.030 ms (2.18% GC)
median time:      498.281 ms (2.63% GC)
mean time:        572.530 ms (16.25% GC)
maximum time:     784.051 ms (36.98% GC)
--------------
samples:          10
evals/sample:     1
```

#### With Views<sup>2</sup>
```
memory estimate:  381.39 MiB
allocs estimate:  118491
--------------
minimum time:     506.433 ms (2.23% GC)
median time:      526.222 ms (2.19% GC)
mean time:        580.181 ms (12.43% GC)
maximum time:     790.616 ms (36.57% GC)
--------------
samples:          9
evals/sample:     1
```

### RGB 512x768
Images used: 'mountainstream` and `lighthouse` from `TestImages`
#### Without Views<sup>1</sup>
```
memory estimate:  566.14 MiB
allocs estimate:  173725
--------------
minimum time:     718.443 ms (2.49% GC)
median time:      719.556 ms (2.51% GC)
mean time:        760.846 ms (7.59% GC)
maximum time:     1.002 s (29.20% GC)
--------------
samples:          7
evals/sample:     1
```

#### With Views<sup>2</sup>
```
memory estimate:  563.15 MiB
allocs estimate:  173737
--------------
minimum time:     808.106 ms (2.39% GC)
median time:      830.977 ms (3.19% GC)
mean time:        916.018 ms (12.60% GC)
maximum time:     1.111 s (27.16% GC)
--------------
samples:          6
evals/sample:     1
```
___
Seems like the without views version performs at least as fast as the version with views. And as the size of the data grows, non-consecutive indexing does take a hit, and copying becomes worth it.
___
1: Code without views:
```(JuliaLang)
x = x[ntuple(i->first(axes(x, i)) : 2 : last(axes(x, i)), ndims(x))...]
ref = ref[ntuple(i->first(axes(ref, i)) : 2 : last(axes(ref, i)), ndims(ref))...]
```
2: Code with views:
```(JuliaLang)
x = @view x[ntuple(i->first(axes(x, i)) : 2 : last(axes(x, i)), ndims(x))...]
ref = @view ref[ntuple(i->first(axes(ref, i)) : 2 : last(axes(ref, i)), ndims(ref))...]
```